Adding Thresholder and RejectOptionClassifier (#997)
fairlearn/fairlearn: docs/user_guide/thresholder.rst

***********
Thresholder
***********

What it does
~~~~~~~~~~~~

:class:`fairlearn.postprocessing.Thresholder` is a post processing technique
which outputs a binary prediction based on group-specific thresholds. The
specified thresholds can work in two directions (positive/negative), and can
be applied to the output of the :code:`predict()` function a regressor,
:code:`predict_proba()[:,1]` of a classifier, or on
:code:`decision_function()` of SVC/LogisticRegression. Examples of these
different applications can be found below. 

What is a threshold?
~~~~~~~~~~~~~~~~~~~~
A threshold is a number which represents the border between positive and
negative predictions. For most classification algorithms, the default
threshold is 0.5, which means that a sample with a probability to get 1
(obtained by calling :code:`estimator.predict_proba()[:,1]`) of 0.6 will be classified
as 1, and a sample with probability to get 1 of 0.4 will be classified as 0.
If this threshold would be changed to 0.7, both samples would be classified
as 0.

Why use it
~~~~~~~~~~~
:class:`Thresholder` is an easy to interpret tool, that can help explore different
trade-offs. Where an algorithm like 

> :class:`fairlearn.postprocessing.ThresholdOptimizer` “blindly” satisfies

Author: When building the docs locally, I don’t experience any problems with this. Does this mean it is okay?

show how to investigate the effect of different thresholds. 

How to use it
~~~~~~~~~~~~~

The group-specific thresholds are specified in :code:`threshold_dict`. This is
a dictionary where the keys are the sensitive feature value of the sub-group,
and the values corresond to the threshold. Below, we will show all different
kind of ways to specify thresholds.

Furthermore, we will plot some information about the obtained predictions,
in order to help gain an insight in the way :class:`Thresholder` works.

Dataset
^^^^^^^^
The dataset used in all examples below, unless specified otherwise, is a
clincial dataset of hospital re-admissions over a ten-year period (1998-2008)
for diabetic patients across 130 different hospitals in the US. Each record
represents the hospital admission records for a patient diagnosed with
diabetes whose stay lasted one to fourteen days. We would like to develop a
classification model, which decides whether the patients should be suggested
to their primary care physicians for an enrollment into the high-risk care
management program. The positive prediction will mean recommendation into the

> care program.

Author: Out of these two options I would prefer the former, mainly because this means that I don’t have to wait for #907, and this can be merged faster. 
To check if I understand correctly, if we would go for this option I should:
-	Move the short description I have here to the dataset user guide
-	Create a separate PR for someone else to complete the description in the dataset user guide
-	Refer to the dataset user guide instead of explaining the dataset in ‘my’ user guide

I should still include the codeblock where I import the dataset and do the train_test_split and stuff in ‘my’ user guide I assume, in order to have the reader understand where all variables later on in the guide are coming from?

.. doctest:: thresholder
    :options:  +NORMALIZE_WHITESPACE

    >>> import pandas as pd
    >>> import numpy as np
    >>> from sklearn.model_selection import train_test_split
    
    >>> #Get dataset, drop columns with unknown sf values
    >>> df = pd.read_csv("https://raw.githubusercontent.com/fairlearn/talks/main/2021_scipy_tutorial/data/diabetic_preprocessed.csv")

    >>> df = df[df['gender'] != 'Unknown/Invalid']
    >>> df = df[df['race'] != 'Unknown']

    >>> Y, A = df.loc[:, 'readmit_30_days'], df.loc[:, 'race']

    >>> #Create dummy's, drop A from X, drop columns that correlate with Y 
    >>> X = pd.get_dummies(df.drop(columns=["race", "discharge_disposition_id",
    ...                                     "readmitted",
    ...                                     "readmit_binary",
    ...                                     "readmit_30_days"]))

    >>> X_train, X_test, Y_train, Y_test, A_train, A_test, df_train, df_test = train_test_split(
    ...        X,
    ...        Y,
    ...        A,
    ...        df,
    ...        test_size=0.50,
    ...        stratify=Y,
    ...        random_state=445)

>     >>> #Resample datset such that training data has same number of positive and negative samples

Reviewer: ```suggestion
    >>> # Resample dataset such that training data has same number of positive and negative samples
```

    
    >>> def resample_dataset(X_train, Y_train, A_train):
    >>>   negative_ids = Y_train[Y_train == 0].index
    >>>   positive_ids = Y_train[Y_train == 1].index
    >>>   np.random.seed(1)
    >>>   balanced_ids = positive_ids.union(np.random.choice(a=negative_ids, size=len(positive_ids)))

    >>>   X_train = X_train.loc[balanced_ids, :]
    >>>   Y_train = Y_train.loc[balanced_ids]
    >>>   A_train = A_train.loc[balanced_ids]
    
    >>>   return X_train, Y_train, A_train

    >>> X_train, Y_train, A_train = resample_dataset(X_train, Y_train, A_train)
   

> Classifier: 'Neutral' predictions

Reviewer: By "neutral", do you mean a uniform threshold for every group? Neutral is somewhat ambiguous...

Author: Yes, that’s what I mean (the standard threshold of 0.5). I changed it to default now, to emphasize that this are predictions without using `Thresholder`. This does not really focus on the fact that there is now a uniform threshold, but I hope the text after the header makes this clear enough?

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The classifier we will be using throughout this example is a
RandomForestClassifier. We first investigate the output when we call predict
on this estimator, without using :class:`Thresholder`. This means all groups
are thresholded at 0.5 (this is the default threshold used by most
classification algorithms). 

Considering the nature of the problem at hand, important aspects of our model
to investigate are:

- How many patients are recommended for care (selection rate = what % of predictions are positive) 
- Reducing false negative rate (we don't want people that need extra care not receiving it)

.. doctest:: thresholder

>     :options:  +NORMALIZE_WHITESPACE

Reviewer: Re: your question:
Yes, this is how you can start code cells that you want executed and where you want the output to be checked. The whitespace option is needed to avoid weird issues with formatting when you output tables. You can remove it and see what happens. It might start complaining... Certainly fine to add it!

    >>> Y_pred_clf = classifier.predict(X_test)

    >>> MetricFrame(metrics=false_negative_rate, 
    ...             y_true=Y_test,
    ...             y_pred=Y_pred_clf, 
    ...             sensitive_features=A_test).by_group
    race
    AfricanAmerican    0.381274
    Asian              0.439024
    Caucasian          0.399399
    Hispanic           0.365217
    Other              0.421875
    Name: false_negative_rate, dtype: object

    >>> plotter.plot_positive_predictions(Y_pred_clf)    

.. figure:: ../auto_examples/images/user_guide_thresholder_no_thresholds.png
    :align: center

It can be useful to also plot the distribution of the probabilities obtained
by calling :code:`predict_proba()[:,1]`. This can give extra insights it what
effect choosing a different threshold will have on the amount of positive
predictions. 

.. doctest:: thresholder
    :options: +NORMALIZE_WHITESPACE

    >>> plotter.plot_proba_distr(Y_pred_proba_clf)

.. figure:: ../auto_examples/images/user_guide_thresholder_proba_distr.png
    :align: center

Two ways of specifying positive thresholds
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> Say, you want to change the threshold of a specific group to 0.4 (all instances
> with a probability to get 1 > 0.4 will be predicted as 1). There are two ways to

Reviewer: ```suggestion
Say, you want to change the threshold of a specific group to 0.4 (all scores
(or probabilities) above 0.4 will be predicted as 1). There are two ways to
```

specify this threshold in the :code:`threshold_dict`: 

>     - 0.4

Reviewer: I'm conflicted as to whether we should allow this...

    - ('>', 0.4)

Changing this threshold for groups 'Asian' and 'Other' shows the impact a
change in threshold of just 0.1 can have on the false negative rates and the
amount of positive predictions.

.. doctest:: thresholder
    :options:  +NORMALIZE_WHITESPACE

    >>> from fairlearn.postprocessing import Thresholder

>     >>> threshold_dict = {'Asian': .4, 'Other': ('>' , .4)}

Reviewer: Aren't these both `.4`?

Also, can we add a leading 0?

    >>> thresholder = Thresholder(estimator=classifier,
    ...                           threshold_dict=threshold_dict,
    ...                           prefit=True,
    ...                           predict_method='predict_proba')
    
    >>> thresholder.fit(X_train,Y_train,sensitive_features=A_train)
    >>> Y_pred_thresholded = thresholder.predict(X_test,sensitive_features=A_test)

    >>> MetricFrame(metrics=false_negative_rate, 
    ...             y_true=Y_test,
    ...             y_pred=Y_pred_thresholded, 
    ...             sensitive_features=A_test).by_group
    race
    AfricanAmerican    0.381274
    Asian              0.121951
    Caucasian          0.399399
    Hispanic           0.365217
    Other                 0.125
    Name: false_negative_rate, dtype: object

    >>> plotter.plot_positive_predictions(Y_pred_thresholded)

.. figure:: ../auto_examples/images/user_guide_thresholder_two_ways_pos_threshold.png
    :align: center

Specifying a negative threshold
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> It is also possible to specify a negative threshold. This means that all
> instances with a probability to get 1 < the specified threshold will be
> predicted as 1. This is done by switching the sign in the threshold. Doing

Reviewer: ```suggestion
It is also possible to specify a negative threshold. This means that all
instances with a score (or probability) less than the specified threshold will be
predicted as 1. This is done by switching the sign in the threshold. Doing
```

this for 'Asian' and 'Other' shows that this flips the false negative rate and
percentage of positive predictions compared to the previous example. Note that
we have to take a threshold just above 0.4 to make sure all the predictions
are switched, as there might be instances with a probability of exaclty 0.4. 

.. doctest:: thresholder
    :options:  +NORMALIZE_WHITESPACE

>     >>> threshold_dict = {'Asian': ('<' , .400001), 'Other': ('<' , .400001)}

Reviewer: These are the same. How come this does anything different at all?
